### DISCUSSION TITLE
High level project planning

### DISCUSSION THREAD

**Discussion by jardinetsouffleton** [corail-research/SeaPearl.jl#245] [Ideas] [answered]
Hello fellow SeaPearl'ers! I think it is time to bring a little -read "a lot of" ;)- structure to this project in order to take it to the next level. To that end, I think we should start by defining some ambitious, but doable milestones. Once we have agreed on those, we will be able to move on and subdivide those into issues, which will make the basis of the tasks we will be working on. I thought a little bit about milestones and I came up with those:

- **Get SeaPearl ready for XCSP**: develop XML model compatibility, ensure that SeaPearl has the right constraints, variable types, etc. to enter the competition. We should also build learned heuristics and run performance tests on instances to validate that our stuff is actually working.
- **MiniCP parity**: Identify constraints, variable types, etc. that are missing from SeaPearl. A little like the above milestone; the 2 could be grouped together
- **AlphaZero-style Monte Carlo Tree Search**: build a functional MCTS + tutorials + related blog posts
- **Variable selection learned heuristic**: build a functional variable selection learned heuristic + tutorials + related blog posts
- **Documentation v1**: formal doc deployed + automated deployment through CI, all tutorials functional and well documented, blog posts explaining the underlying theory. 
- **Internal/project standards documentation**: Write a setup guide for new contributors -especially those new to julialang-, contribution guidelines, coding standards, issue/bug templates

There it is. I think there is a lot more to be added as milestones and I was hoping that @qcappart and @3rdCore would chip in so that we have a more complete view of exactly what needs to be built. Of course, anyone is free to contribute to this thread and I welcome any disagreement or comment!

Let's make SeaPearl a top-notch solver!!!

**Comment by qcappart** [comment, marked as answer]
Hello ! Thank you for curating this list ! I agree with the milestones. 

The following tasks are doable in the short term ( < 1 year) and could be performed before the XCSP contest.

- Get SeaPearl ready for XCSP
- MiniCP parity
- Documentation v1
- Internal/project standards documentation

In any case, completing these tasks will be greatly beneficial for SeaPearl.

The tasks _MCTS and variable selection heuristics_,  are more to fundamental research questions.

**Reply by jardinetsouffleton** [reply]
Perfect, I will start by creating milestones for these items. Let's discuss the other points in person. @3rdCore would you see anything else on the roadmap?

**Reply by 3rdCore** [reply]
Hello, 
Nice job ! I don't see anything to add regarding the introduction materials (documentation, readme, blog post...). 

At a more technical level, I'd also consider the following task : 
- **NN pipeline computational improvement:** there is certainly some work to do to improve operations done to compute the forward (and backward) pass on both the encoder and the decoder part.
- **Representation matters for new Variable and/or Constraint:**  this is a more fundamental research question but each new Variable and Constraint that we will implement (regarding the XCSP requirement for example) must come with a way to fully encode it using our representation, otherwise we wouldn't be able to use our RL tools to speed up solving process. 

Also, I think that some improvement could be done on SeaPearlZoo. Should we discuss them here or on a separate thread?
